Source/Basetypes/BigFloat.cs

```diff
@@ -523,73 +434,158 @@ public void FloorCeiling(out BIM floor, out BIM ceiling)
+        bool ySignBitSet = y.value == "" ? y.isSignBitSet : y.value[0] == '-';
+        return new BigFloat((xSignBitSet ^ ySignBitSet ? "-" : "+") + "oo", x.significandSize, x.exponentSize);
+      }
 
-    public bool IsSpecialType {
-      get {
-        if (value == "")
-          return false;
-        return (value.Equals("NaN") || value.Equals("+oo") || value.Equals("-oo"));
+      BIM xsig = x.significand, ysig = y.significand;
+      BIM xexp = x.exponent, yexp = y.exponent;
+
+      BIM hiddenBitPow = BIM.Pow(2, x.significandSize - 1);
+
+      if (xexp > 0) {
+        xsig += hiddenBitPow;
+      } else {
+        ++xexp;
       }
-    }
 
-    public bool IsPositive {
-      get {
-        return !IsNegative;
+      if (yexp > 0) {
+        ysig += hiddenBitPow;
+      } else {
+        ++yexp;
       }
+
+      ysig *= xsig;
+      yexp += xexp - (BIM.Pow(2, x.exponentSize - 1) - 1) - (x.significandSize - 1);
+
+      while (ysig >= hiddenBitPow * 2 || yexp <= 0) {
+        ysig >>= 1;
+        ++yexp;
+      }
+
+      while (ysig < hiddenBitPow && yexp > 1) {
+        ysig <<= 1;
+        --yexp;
+      }
+
+      if (ysig < hiddenBitPow) {
+        yexp = 0;
+      } else {
+        ysig -= hiddenBitPow;
+      }
+
+      if (yexp >= BIM.Pow(2, x.exponentSize) - 1) {
+        return new BigFloat(x.isSignBitSet ^ y.isSignBitSet ? "-oo" : "+oo", x.significandSize, x.exponentSize);
+      }
+
+      return new BigFloat(x.isSignBitSet ^ y.isSignBitSet, ysig, yexp, x.significandSize, x.exponentSize);
     }
 
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Some basic comparison operations
+
     public bool IsZero {
       get {
-        return significand.Equals(BigNum.ZERO) && Exponent == BIM.Zero;
+        return value == "" && significand.IsZero && exponent.IsZero;
       }
     }
 
     [Pure]
     public int CompareTo(BigFloat that) {
-      if (this.exponent > that.exponent)
-        return 1;
-      if (this.exponent < that.exponent)
-        return -1;
-      if (this.significand == that.significand)
+      Contract.Requires(exponentSize == that.exponentSize);
+      Contract.Requires(significandSize == that.significandSize);
+
+      if (value == "" && that.value == "") {
+        int cmpThis = IsZero ? 0 : isSignBitSet ? -1 : 1;
+        int cmpThat = that.IsZero ? 0 : that.isSignBitSet ? -1 : 1;
+
+        if (cmpThis == cmpThat) {
+          if (exponent == that.exponent) {
+            return cmpThis * significand.CompareTo(that.significand);
+          }
+
+          return cmpThis * exponent.CompareTo(that.exponent);
+        }
+
+        if (cmpThis == 0) {
+          return -cmpThat;
+        }
+
+        return cmpThis;
+      }
+
+      if (value == that.value) {
         return 0;
-      return this.significand > that.significand ? 1 : -1;
+      }
+
+      if (value == "NaN" || that.value == "+oo" || value == "-oo" && that.value != "NaN") {
```

Comment:
Wait, what if `that.value == "NaN"`?

Reply:
If we've reached this if statement and that.value == "NaN", then we should return 1. All values, including -infinity, are greater than NaN in the CompareTo method. However, note that the expression `-infinity > NaN` should return false. See my other comment for more details.